fdbclient/SystemData.cpp

```diff
@@ -43,20 +45,69 @@ const Key keyServersKey( const KeyRef& k ) {
 const KeyRef keyServersKey( const KeyRef& k, Arena& arena ) {
 	return k.withPrefix( keyServersPrefix, arena );
 }
-const Value keyServersValue( const vector<UID>& src, const vector<UID>& dest ) {
+const Value keyServersValue( Standalone<RangeResultRef> result, const std::vector<UID>& src, const std::vector<UID>& dest ) {
+	std::vector<Tag> srcTag;
+	std::vector<Tag> destTag;
+
+	for (const KeyValueRef kv : result) {
+		UID uid = decodeServerTagKey(kv.key);
+		if (std::find(src.begin(), src.end(), uid) != src.end()) {
+			srcTag.push_back( decodeServerTagValue(kv.value) );
+		}
+		if (std::find(dest.begin(), dest.end(), uid) != dest.end()) {
+			destTag.push_back( decodeServerTagValue(kv.value) );
+		}
+	}
+
+	return keyServersValue(srcTag, destTag);
+}
+const Value keyServersValue( const std::vector<Tag>& srcTag, const std::vector<Tag>& destTag ) {
 	// src and dest are expected to be sorted
-	ASSERT( std::is_sorted(src.begin(), src.end()) && std::is_sorted(dest.begin(), dest.end()) );
-	BinaryWriter wr((IncludeVersion())); wr << src << dest;
+	BinaryWriter wr(IncludeVersion()); wr << srcTag << destTag;
 	return wr.toValue();
 }
-void decodeKeyServersValue( const ValueRef& value, vector<UID>& src, vector<UID>& dest ) {
-	if (value.size()) {
-		BinaryReader rd(value, IncludeVersion());
-		rd >> src >> dest;
-	} else {
+
+void decodeKeyServersValue( Standalone<RangeResultRef> result, const ValueRef& value,
+                            std::vector<UID>& src, std::vector<UID>& dest  ) {
+	if (value.size() == 0) {
 		src.clear();
 		dest.clear();
+		return;
+	}
+
+	BinaryReader rd(value, IncludeVersion());
+	rd.checkpoint();
+	int srcLen, destLen;
+	rd >> srcLen;
+	rd.readBytes(srcLen * sizeof(Tag));
+	rd >> destLen;
+	rd.rewind();
+
+	if (value.size() == sizeof(ProtocolVersion) + sizeof(int) + srcLen * sizeof(UID) + sizeof(int) + destLen * sizeof(UID)) {
+		rd >> src >> dest;
+		return;
+	}
+
+	// If this is not true, then our math was wrong.
+	ASSERT(value.size() == sizeof(ProtocolVersion) + sizeof(int) + srcLen * sizeof(Tag) + sizeof(int) + destLen * sizeof(Tag));
+
+	std::vector<Tag> srcTag, destTag;
+	rd >> srcTag >> destTag;
+
+	src.clear();
+	dest.clear();
+
+	for (const KeyValueRef kv : result) {
+		Tag tag = decodeServerTagValue(kv.value);
+		if (std::find(srcTag.begin(), srcTag.end(), tag) != srcTag.end()) {
```

Comment:
Since this func is on hot path and the keyServer list (i.e., result) can be very long (I think the length can be 100K. correct me if I'm wrong.), adding a counter inside the for() loop, increase it by 1 every time we add it to `src` or `dest`  and break the loop once the counter = `srcTag.size() + destTag.size()`. 
This may speed up the search a bit.